tests/unit/src/test/scala/tests/ReferenceLspSuite.scala

```diff
@@ -107,6 +107,113 @@ class ReferenceLspSuite extends BaseRangesSuite("reference") {
     } yield ()
   }
 
+  /** References should be found correctly regardless of the whether or not
+   *  the declaring code and the refering share the same compilation unit.
+   *
+   *  This checks both cases by first placing them in the same file and then
+   *  spread across different files. The package stays the same - this is
+   *  actually necessary to test that references to synthetics can be found
+   *  even in the absence of an explicit `import`, which is often the case when
+   *  the referencing code shares with the declaring code the same package
+   *  (but possibly a different compilation unit)
+   */
+  def checkInSamePackage(
+      name: String,
+      code: String,
+      moreCode: String*
+  ): Unit = {
+    def input(chunks: Seq[String]): String =
+      chunks.zipWithIndex
+        .map {
+          case (chunk, i) =>
+            s"""|/a/src/main/scala/a/Chunk$i.scala
+                |package a
+                |$chunk
+                |""".stripMargin
+        }
+        .mkString("\n")
+    check(s"$name-together", input(Seq((code +: moreCode) mkString "\n")))
+    check(s"$name-split-up", input(code +: moreCode))
+  }
+
+  checkInSamePackage(
+    "simple-case-class",
+    """|case class <<Ma@@in>>(name: String)
+       |object F {
+       |  val ma = <<Main>>("a")
+       |}
+       |""".stripMargin,
+    """|object Other {
+       |  val mb = <<Main>>.apply("b")
+       |}
+       |""".stripMargin
+  )
+
+  checkInSamePackage( // FIXME: doesn't find the case class declaration
```

Comment:
Let's maybe add the reference to https://github.com/scalameta/metals/issues/1553

Reply:
OK, I'm going to reference the conversation where it becomes clear that we've got 2 bugs (and what the bugs are).